collect_app/src/main/java/org/odk/collect/android/activities/FormFillingActivity.java

```diff
@@ -556,15 +556,7 @@ private void setupViewModels(FormEntryViewModelFactory formEntryViewModelFactory
             }
             ValidationResult validationResult = consumable.getValue();
             if (validationResult instanceof FailedValidationResult failedValidationResult) {
-                try {
-                    createConstraintToast(failedValidationResult.getIndex(), failedValidationResult.getStatus());
-                    if (getFormController().indexIsInFieldList() && getFormController().getQuestionPrompts().length > 1) {
-                        getCurrentViewIfODKView().highlightWidget(failedValidationResult.getIndex());
-                    }
-                } catch (RepeatsInFieldListException e) {
-                    createErrorDialog(new FormError.NonFatal(e.getMessage()));
-                }
-
+                getCurrentViewIfODKView().setErrorForQuestionWithIndex(failedValidationResult.getIndex(), formEntryViewModel.getErrorMessage(failedValidationResult, this));
```

Comment:
I'm generally really liking the structure of all this, but before I get into all the details: I was wondering if it would make sense to have the error message be part of the `FailedValidationResult` rather than having to reach back in the `FormEntryViewModel` to get it here?

Reply:
Not really. The only place where we construct `FailedValidationResult` is `JavaRosaFormController` because that's where the validation takes place. On one hand, it would be good to move the method responsible for generating error messages there since first that method calls some javarosa methods to check if there is a custom error message but if not it needs to read our strings, and to do that we need `context` that would need to be passed through a few methods and that wouldn't be easy.

Reply:
Ah good point. I still think it'd be better to have the observation not have to reach back into the `FormEntryViewModel` to simplify the amount of interactions here. Maybe instead of `getValidationResult` returning the JavaRosa type, it could return a `Result<Unit, String>?` so the `FormFillingActivity` side would look like:

```kotlin
formEntryViewModel.getValidationError().observe(this, consumable -> {
    if (consumable.isConsumed()) {
        return;
    }
    String result = consumable.getValue();
    if (result instance Result.Error) {
        getCurrentViewIfODKView().setErrorForQuestionWithIndex(failedValidationResult.getIndex(), ((Result.Error) result.value));
        swipeHandler.setBeenSwiped(false);
    } else if (result insanceof Result.Success) {
        SnackbarUtils.showLongSnackbar(findViewById(R.id.llParent), getString(org.odk.collect.strings.R.string.success_form_validation), findViewById(R.id.buttonholder));
    }

    consumable.consume();
});
```

That way the `FormEntryViewModel` API is smaller, and we don't expose JavaRosa types from it.

Reply:
but I need `FormIndex` here. In your example, it's `failedValidationResult.getIndex()` but `failedValidationResult` does not exist anymore. I would rather need a wrapper around `ValidationResult` that would add the message and then I would be able to use it in a similar way as you described. Or maybe I'm missing something?

Reply:
> I would rather need a wrapper around ValidationResult that would add the message and then I would be able to use it in a similar way as you described. Or maybe I'm missing something?

Ah sorry that was a mistake on my part. I like the idea of a wrapper that `FormEntryViewModel` owns though that gives you a message and an index.

Reply:
Ah it would still require passing context through many different methods to be able to read strings. I gave it a try and I think it doesn't make sense. In the view model class I would need to edit:
- `public void moveForward(HashMap<FormIndex, IAnswerData> answers)`
- `public void moveForward(HashMap<FormIndex, IAnswerData> answers, Boolean evaluateConstraints) `
- `public void moveBackward(HashMap<FormIndex, IAnswerData> answers)`
- `public boolean updateAnswersForScreen(HashMap<FormIndex, IAnswerData> answers, Boolean evaluateConstraints)`
- `public boolean saveScreenAnswersToFormController(HashMap<FormIndex, IAnswerData> answers, Boolean evaluateConstraints)`
- `public void validate()
`

Reply:
Ok I've implemented something like that.